cmd/config_consolidation_test.go

```diff
@@ -375,6 +382,86 @@ func getConfigConsolidationTestCases() []configConsolidationTestCase {
+			}, c.Options.DNS)
+		}},
+		{opts{env: []string{"K6_DNS=ttl=5,select=round-robin"}}, exp{}, func(t *testing.T, c Config) {
+			assert.Equal(t, lib.DNSConfig{
+				TTL:    null.StringFrom("5"),
+				Select: lib.NullDNSSelect{DNSSelect: lib.DNSRoundRobin, Valid: true},
+				Policy: lib.NullDNSPolicy{DNSPolicy: lib.DNSpreferIPv4, Valid: false},
+			}, c.Options.DNS)
+		}},
+		{opts{env: []string{"K6_DNS=ttl=inf,select=random,policy=preferIPv6"}}, exp{}, func(t *testing.T, c Config) {
+			assert.Equal(t, lib.DNSConfig{
+				TTL:    null.StringFrom("inf"),
+				Select: lib.NullDNSSelect{DNSSelect: lib.DNSRandom, Valid: true},
+				Policy: lib.NullDNSPolicy{DNSPolicy: lib.DNSpreferIPv6, Valid: true},
+			}, c.Options.DNS)
+		}},
+		// This is functionally invalid, but will error out in validation done in js.parseTTL().
+		{opts{cli: []string{"--dns", "ttl=-1"}}, exp{}, func(t *testing.T, c Config) {
+			assert.Equal(t, lib.DNSConfig{
+				TTL:    null.StringFrom("-1"),
+				Select: lib.NullDNSSelect{DNSSelect: lib.DNSRandom, Valid: false},
+				Policy: lib.NullDNSPolicy{DNSPolicy: lib.DNSpreferIPv4, Valid: false},
+			}, c.Options.DNS)
+		}},
+		{opts{cli: []string{"--dns", "ttl=0,blah=nope"}}, exp{cliReadError: true}, nil},
+		{opts{cli: []string{"--dns", "ttl=0"}}, exp{}, func(t *testing.T, c Config) {
+			assert.Equal(t, lib.DNSConfig{
+				TTL:    null.StringFrom("0"),
+				Select: lib.NullDNSSelect{DNSSelect: lib.DNSRandom, Valid: false},
+				Policy: lib.NullDNSPolicy{DNSPolicy: lib.DNSpreferIPv4, Valid: false},
+			}, c.Options.DNS)
+		}},
+		{opts{cli: []string{"--dns", "ttl=5s,select="}}, exp{cliReadError: true}, nil},
+		{
+			opts{fs: defaultConfig(`{"dns": {"ttl": "0", "select": "round-robin", "policy": "onlyIPv4"}}`)},
+			exp{},
+			func(t *testing.T, c Config) {
+				assert.Equal(t, lib.DNSConfig{
+					TTL:    null.StringFrom("0"),
+					Select: lib.NullDNSSelect{DNSSelect: lib.DNSRoundRobin, Valid: true},
+					Policy: lib.NullDNSPolicy{DNSPolicy: lib.DNSonlyIPv4, Valid: true},
+				}, c.Options.DNS)
+			},
+		},
+		{
+			opts{
+				fs:  defaultConfig(`{"dns": {"ttl": "0"}}`),
+				env: []string{"K6_DNS=ttl=30,policy=any"},
+			},
+			exp{},
+			func(t *testing.T, c Config) {
+				assert.Equal(t, lib.DNSConfig{
+					TTL:    null.StringFrom("30"),
+					Select: lib.NullDNSSelect{DNSSelect: lib.DNSRandom, Valid: false},
+					Policy: lib.NullDNSPolicy{DNSPolicy: lib.DNSany, Valid: true},
+				}, c.Options.DNS)
+			},
+		},
+		{
+			// CLI overrides all, falling back to env
+			opts{
+				fs:  defaultConfig(`{"dns": {"ttl": "60", "select": "first"}}`),
+				env: []string{"K6_DNS=ttl=30,select=random,policy=any"},
+				cli: []string{"--dns", "ttl=5"},
+			},
+			exp{},
+			func(t *testing.T, c Config) {
+				assert.Equal(t, lib.DNSConfig{
+					TTL:    null.StringFrom("5"),
+					Select: lib.NullDNSSelect{DNSSelect: lib.DNSRandom, Valid: true},
+					Policy: lib.NullDNSPolicy{DNSPolicy: lib.DNSany, Valid: true},
+				}, c.Options.DNS)
```

Comment:
Yeah, I can see how it could be confusing. But still think the flexibility outweighs any chance of confusion. It would be useful to e.g. have a base configuration in the script, but then be able to override a single sub-option via the env or CLI, instead of having to redefine everything.

Besides, the only reason `dns` is composite is to avoid adding a bunch of top-level options to our messy config system, but otherwise having them separated with a prefix (e.g. `--dns-ttl`, `--dns-policy`, etc.) would be more preferable IMO. These would also map cleanly to env vars as `K6_DNS_TTL`, etc., if we had a sane config system.